File: lib/models/musical.dart
class Musical {
  final int id;
  final String title;
  final String cast;
  final String startDate;
  final String endDate;
  final String runtime;
  final String theater;
  final String posterUrl;
  final String? period;
  final double? interparkRating;
  final double? yes24Rating;
  final String? interparkUrl;
  final String? yes24Url;

  Musical({
    required this.id,
    required this.title,
    required this.cast,
    required this.startDate,
    required this.endDate,
    required this.runtime,
    required this.theater,
    required this.posterUrl,
    this.period,
    this.interparkRating,
    this.yes24Rating,
    this.interparkUrl,
    this.yes24Url,
  });

  factory Musical.fromJson(Map<String, dynamic> json) {
    // 1) JSON 키가 posterUrl 또는 poster_url 중 하나일 수 있으니 둘 다 시도
    final raw = (json['posterUrl'] ?? json['poster_url'] ?? '') as String;
    // 2) 만약 //로 시작하면 https: 스킴 붙여 주기
    final fixedUrl = raw.startsWith('//') ? 'https:$raw' : raw;
    return Musical(
      id: json['id'] as int,
      title: json['title'] as String,
      cast: json['cast'] as String? ?? '',
      startDate: json['startDate'] as String? ?? '',
      endDate: json['endDate'] as String? ?? '',
      runtime: json['runtime'] as String? ?? '',
      theater: json['theater'] as String? ?? '',
      posterUrl: fixedUrl,
      period: json['period'] as String?,
      interparkRating: json['interparkRating'] != null ? (json['interparkRating'] as num).toDouble() : null,
      yes24Rating: json['yes24Rating'] != null ? (json['yes24Rating'] as num).toDouble() : null,
      interparkUrl: json['interparkUrl'] as String?,
      yes24Url: json['yes24Url'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'title': title,
    'cast': cast,
    'startDate': startDate,
    'endDate': endDate,
    'runtime': runtime,
    'theater': theater,
    'poster_url': posterUrl,
    'period': period,
    'interparkRating': interparkRating,
    'yes24Rating': yes24Rating,
    'interparkUrl': interparkUrl,
    'yes24Url': yes24Url,
  };
}

File: src/main/java/com/melodical/backend/entity/Musical.java
package com.melodical.backend.entity;

import jakarta.persistence.*;
import lombok.*;
import java.time.LocalDateTime;

@Entity
@Table(name = "musical")
@Getter @Setter @NoArgsConstructor @AllArgsConstructor @Builder
public class Musical {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "interpark_id", unique = true)
    private String interparkId; // Interpark URL의 마지막 8자리 숫자

    @Column(name = "interpark_url")
    private String interparkUrl; // Interpark 티켓 페이지 URL

    @Column(name = "yes24_url")
    private String yes24Url; // Yes24 티켓 페이지 URL

    private String cast;
    private String endDate;
    private String posterUrl;
    private String runtime;
    private String startDate;
    private String theater;
    private String title;

    // 추천 시스템을 위한 추가 메타데이터
    @Column(name = "genre")
    private String genre; // 뮤지컬 장르 (예: "로맨스", "코미디", "드라마" 등)

    @Column(name = "tags")
    private String tags; // 태그들 (콤마로 구분, 예: "가족뮤지컬,클래식,브로드웨이")

    @Column(name = "series_id")
    private String seriesId; // 시리즈 식별자 (같은 작품의 다른 시즌/버전)

    @Column(name = "producer")
    private String producer; // 제작사

    @Column(name = "age_rating")
    private String ageRating; // 연령 등급 (전체관람가, 12세이상 등)

    @Column(name = "price_min")
    private Integer priceMin; // 최저가격

    @Column(name = "price_max")
    private Integer priceMax; // 최고가격

    @Column(name = "region")
    private String region; // 공연 지역 (서울, 부산 등)

    @Column(name = "is_on_sale")
    private Boolean isOnSale; // 현재 판매중 여부

    @Column(name = "is_new")
    private Boolean isNew; // 신작 여부

    @Column(name = "popularity_score")
    private Double popularityScore; // 인기 점수 (조회수, 예매율 기반)

    @Column(name = "view_count")
    private Long viewCount; // 조회 수

    @Column(name = "booking_rate")
    private Double bookingRate; // 예매율

    @Column(name = "created_at")
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
        if (viewCount == null) viewCount = 0L;
        if (popularityScore == null) popularityScore = 0.0;
        if (bookingRate == null) bookingRate = 0.0;
        if (isOnSale == null) isOnSale = false;
        if (isNew == null) isNew = false;
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}
